Event Timing: assign interactionID to keypress

francovs opened the pull request:
#### 27ebefc6f80d92f5e938e92ebd585712bb657f2f
<pre>
Event Timing: assign interactionID to keypress
<a href="https://bugs.webkit.org/show_bug.cgi?id=298904">https://bugs.webkit.org/show_bug.cgi?id=298904</a>
<a href="https://rdar.apple.com/160647204">rdar://160647204</a>

Reviewed by Ryosuke Niwa.

Assigns an interactionID to keypress events when they are
part of a keydown-keypress sequence. This is not in
accordance to the event timing spec, but is required by
one WPT test (event-timing/interactionid-keypress.html).

Implementing this required deferring the dispatch of keypress
events until the keyup event happens, in a similar manner to
what was already happening for keydown (see the change on
m_pendingKeyDowns).

Matching a keypress to a keydown required an extra function
(KeyboardEvent::keyCodeForKeyDown()) and some change in logic
because the keyCode of keypresses is normally overridden. We now
preserve the original value and do the overriding in
windowsVirtualKeyCode().

Now passing:
* LayoutTests/imported/w3c/web-platform-tests/event-timing/interactionid-keypress.html

Canonical link: <a href="https://commits.webkit.org/300198@main">https://commits.webkit.org/300198@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/45ce7178f32fea679f8038fc8804c3d4a3db3c82

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/121591 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/41288 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/31947 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/128126 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/73680 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/123467 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/42006 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/49867 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/92411 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/61450 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/124543 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/33529 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/108919 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/73073 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/32550 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/27080 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/71624 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/103032 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/27248 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/130879 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/48519 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/36924 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/100967 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/48887 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/105131 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/100853 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25587 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/46253 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/24342 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/45221 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/48378 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/54090 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/47855 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/51200 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/49533 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/dom/KeyboardEvent.cpp:
```diff
@@ -208,6 +208,11 @@ int KeyboardEvent::keyCode() const
     return charCode();
 }
 
+int KeyboardEvent::keyCodeAsKeyDown() const
```
I'd call this `keyCodeForKeyDown` instead.

rniwa commented on Source/WebCore/platform/PlatformKeyboardEvent.h:
```diff
@@ -160,6 +162,7 @@ namespace WebCore {
         String m_code;
         String m_keyIdentifier;
         int m_windowsVirtualKeyCode { 0 };
+        int m_unmodifiedWindowsVirtualKeyCode { 0 };
```
This doesn't tell us how "unmodified" this value is. Given the only difference is from whether input method overrode the key-code to be 229 or not, might be more straight forward to add a boolean indicating whether input-method-override is in effect or not, and then return 229 in `windowsVirtualKeyCode()` when that boolean flag is set. This will avoid having to duplicate this key-code information twice.

francovs replied:
It's not just IME: `PlatformKeyboardEvent::disambiguateKeyDownEvent()` does `m_windowsVirtualKeyCode = 0;` for keypress events (except on Windows), and this is what makes matching keypresses to keydowns hard. Unfortunately, the spec requires this inconsistency: https://www.w3.org/TR/uievents/#determine-keypress-keyCode

Still, point taken: adding an extra check to `windowsVirtualKeyCode()` made `m_unmodifiedWindowsVirtualKeyCode` unnecessary.

rniwa commented on Source/WebCore/dom/KeyboardEvent.cpp:
```diff
@@ -208,6 +208,11 @@ int KeyboardEvent::keyCode() const
 
+int KeyboardEvent::keyCodeAsKeyDown() const
+{
+    return windowsVirtualKeyCodeWithoutLocation(m_underlyingPlatformEvent->unmodifiedWindowsVirtualKeyCode());
```
We should probably assert that `m_underlyingPlatformEvent` is in fact keydown.

francovs replied:
It shouldn't be a keydown! This is supposed to make a keypress event behave as if it were a keydown, so that the key codes can be matched.

Added `ASSERT(type() == eventNames().keypressEvent);` to clarify the intent.

rniwa replied:
no, I meant that we should type-check `m_underlyingPlatformEvent`. `type()` is `keypress` but `m_underlyingPlatformEvent->type()` is `keydown`, right?

francovs replied:
No, I think you are assuming a keypress has a keydown as its underlying event, and mixing up `Event::m_underlyingEvent` and `KeyboardEvent::m_underlyingPlatformEvent`. This isn't the case, but it makes a lot more sense than what is on the PR now. ~Will refactor!~ (EDIT: refactor didn't work on Windows because it does not derive keypresses from keydowns 😞)

Currently, `m_underlyingPlatformEvent.m_type` (there is no `.type()`) would be `PlatformEventType::Char`, which also corresponds to a keypress.

rniwa replied:
oh sorry, I misread.

francovs committed:
Event Timing: assign interactionID to keypress
https://bugs.webkit.org/show_bug.cgi?id=298904
rdar://160647204

Reviewed by Ryosuke Niwa.

Assigns an interactionID to keypress events when they are
part of a keydown-keypress sequence. This is not in
accordance to the event timing spec, but is required by
one WPT test (event-timing/interactionid-keypress.html).

Implementing this required deferring the dispatch of keypress
events until the keyup event happens, in a similar manner to
what was already happening for keydown (see the change on
m_pendingKeyDowns).

Matching a keypress to a keydown required an extra function
(KeyboardEvent::keyCodeForKeyDown()) and some change in logic
because the keyCode of keypresses is normally overridden. We now
preserve the original value and do the overriding in
windowsVirtualKeyCode().

Now passing:
* LayoutTests/imported/w3c/web-platform-tests/event-timing/interactionid-keypress.html

Canonical link: https://commits.webkit.org/300198@main
